src/replay/service.zig

```diff
@@ -38,162 +41,438 @@ const MAX_ENTRIES: u64 = 1024 * 1024; // 1 million slots is about 5 days
-    progress_map: *ProgressMap,
-    ledger_db: LedgerDB,
-    execution: ReplayExecutionState,
+    progress_map: ProgressMap,
+    ledger: LedgerRef,
+
+    fork_choice: HeaviestSubtreeForkChoice,
+    replay_tower: sig.consensus.ReplayTower,
+    latest_validator_votes: LatestValidatorVotes,
+    status_cache: sig.core.StatusCache,
+    slot_data: replay.edge_cases.SlotData,
+
+    arena_state: std.heap.ArenaAllocator.State,
+
+    senders: Senders,
+    receivers: Receivers,
+
+    fn deinit(self: *ReplayState) void {
+        self.thread_pool.shutdown();
+        self.thread_pool.deinit();
+
+        self.slot_tracker.deinit(self.allocator);
+
+        self.epochs.deinit(self.allocator);
+
+        self.progress_map.deinit(self.allocator);
+
+        self.fork_choice.deinit();
+        self.latest_validator_votes.deinit(self.allocator);
+        self.slot_data.deinit(self.allocator);
+    }
 
     fn init(deps: ReplayDependencies) !ReplayState {
-        const thread_pool = try deps.allocator.create(ThreadPool);
-        errdefer deps.allocator.destroy(thread_pool);
-        thread_pool.* = ThreadPool.init(.{ .max_threads = NUM_THREADS });
-
-        var root_slot_state = deps.root_slot_state;
-        const last_blockhash = root_slot_state.blockhash_queue.readField("last_hash") orelse
-            return error.InvalidBlockhashQueue;
-
-        const slot_tracker = try deps.allocator.create(SlotTracker);
-        errdefer deps.allocator.destroy(slot_tracker);
-        slot_tracker.* = try .init(deps.allocator, deps.root_slot, .{
-            .constants = deps.root_slot_constants,
-            .state = root_slot_state,
+        var slot_tracker: SlotTracker = try .init(deps.allocator, deps.root.slot, .{
+            .constants = deps.root.constants,
+            .state = deps.root.state,
         });
         errdefer slot_tracker.deinit(deps.allocator);
 
-        const epoch_tracker = try deps.allocator.create(EpochTracker);
-        errdefer deps.allocator.destroy(epoch_tracker);
-        epoch_tracker.* = .{ .schedule = deps.epoch_schedule };
+        var epoch_tracker: EpochTracker = .{ .schedule = deps.epoch_schedule };
         errdefer epoch_tracker.deinit(deps.allocator);
+
         try epoch_tracker.epochs
             .put(deps.allocator, deps.current_epoch, deps.current_epoch_constants);
 
-        const progress_map = try deps.allocator.create(ProgressMap);
-        progress_map.* = ProgressMap.INIT;
-        try progress_map.map.put(
+        const progress_map, const fork_choice = try initProgressAndForkChoiceWithLockedSlotForks(
+            deps.allocator,
+            deps.logger,
+            &slot_tracker,
+            &epoch_tracker,
+            deps.my_identity,
+            deps.vote_identity,
+            deps.ledger.reader.*,
+        );
+        errdefer progress_map.deinit(deps.allocator);
+        errdefer fork_choice.deinit();
+
+        // TODO(ink): in agave replay_tower isn't created directly in replay,
```

Comment:
This is not really a TODO because there's nothing we actually need to do. It's just an observation that agave has different requirements and a different design. I'm sure a lot of data is going to need to be shuffled around as sig's requirements evolve, but I don't think we can reasonably dictate what actually needs to be done at this point. It's helpful to have some notes in the code for this, but it's not really an actionable task at this time. I would change this to NOTE or unqualified.

Reply:
[519e304](https://github.com/Syndica/sig/pull/810/commits/519e3047d02aa8e44a289b93c8b54dd8a8941668)